shared-bindings/displayio/EPaperDisplay.c

```diff
@@ -0,0 +1,313 @@
+        if (displays[i].display.base.type == NULL ||
+            displays[i].display.base.type == &mp_type_NoneType) {
+            self = &displays[i].epaper_display;
+            break;
+        }
+    }
+    if (self == NULL) {
+        mp_raise_RuntimeError(translate("Too many displays"));
+    }
+
+    mp_float_t refresh_time = mp_obj_get_float(args[ARG_refresh_time].u_obj);
+    mp_float_t seconds_per_frame = mp_obj_get_float(args[ARG_seconds_per_frame].u_obj);
+
+    mp_int_t write_color_ram_command = NO_COMMAND;
+    mp_int_t highlight_color = args[ARG_highlight_color].u_int;
+    if (args[ARG_write_color_ram_command].u_obj != mp_const_none) {
+        write_color_ram_command = mp_obj_get_int(args[ARG_write_color_ram_command].u_obj);
+    }
+
+    self->base.type = &displayio_epaperdisplay_type;
+    common_hal_displayio_epaperdisplay_construct(
+        self,
+        display_bus,
+        start_bufinfo.buf, start_bufinfo.len, stop_bufinfo.buf, stop_bufinfo.len,
+        args[ARG_width].u_int, args[ARG_height].u_int, args[ARG_ram_width].u_int, args[ARG_ram_height].u_int, args[ARG_colstart].u_int, args[ARG_rowstart].u_int, rotation,
+        args[ARG_set_column_window_command].u_int, args[ARG_set_row_window_command].u_int,
+        args[ARG_set_current_column_command].u_int, args[ARG_set_current_row_command].u_int,
+        args[ARG_write_black_ram_command].u_int, args[ARG_black_bits_inverted].u_bool, write_color_ram_command, args[ARG_color_bits_inverted].u_bool, highlight_color, args[ARG_refresh_display_command].u_int, refresh_time,
+        busy_pin, args[ARG_busy_state].u_bool, seconds_per_frame, args[ARG_always_toggle_chip_select].u_bool
+        );
+
+    return self;
+}
+
+// Helper to ensure we have the native super class instead of a subclass.
+static displayio_epaperdisplay_obj_t* native_display(mp_obj_t display_obj) {
+    mp_obj_t native_display = mp_instance_cast_to_native_base(display_obj, &displayio_epaperdisplay_type);
+    mp_obj_assert_native_inited(native_display);
+    return MP_OBJ_TO_PTR(native_display);
+}
+
+//|   .. method:: show(group)
+//|
+//|     Switches to displaying the given group of layers. When group is None, the default
+//|     CircuitPython terminal will be shown.
+//|
+//|     :param Group group: The group to show.
+STATIC mp_obj_t displayio_epaperdisplay_obj_show(mp_obj_t self_in, mp_obj_t group_in) {
+    displayio_epaperdisplay_obj_t *self = native_display(self_in);
+    displayio_group_t* group = NULL;
+    if (group_in != mp_const_none) {
+        group = MP_OBJ_TO_PTR(native_group(group_in));
+    }
+
+    bool ok = common_hal_displayio_epaperdisplay_show(self, group);
+    if (!ok) {
+        mp_raise_ValueError(translate("Group already used"));
+    }
+    return mp_const_none;
+}
+MP_DEFINE_CONST_FUN_OBJ_2(displayio_epaperdisplay_show_obj, displayio_epaperdisplay_obj_show);
+
+//|   .. method:: refresh()
+//|
+//|     Refreshes the display immediately or raises an exception if too soon. Use
```

Comment:
I chose to do this because one must be very careful when to call refresh with epaper displays because the next refresh is minutes away. It needs to be a deliberate action and should fail hard if not done correctly.

Reply:
I just mean that if you call refresh() too soon, it will do nothing (and return False), as opposed to forcing a refresh. That way you don't have to keep track of the time so precisely. E.g. you could call refresh once a minute to update a clock, and it wouldn't actually get updated for 3 minutes. Otherwise you have to know when 3 minutes is up.